src/mesh/difops.cxx

```diff
@@ -87,73 +87,83 @@ const Field3D Grad_parP(const Field3D &apar, const Field3D &f) {
 
   Field3D gys(mesh);
   gys.allocate();
+  gys.setLocation(f.getLocation());
 
   // Need Y derivative everywhere
-  for(int x=1;x<=mesh->LocalNx-2;x++)
-    for(int y=1;y<=mesh->LocalNy-2;y++)
-      for(int z=0;z<ncz;z++) {
-        gys(x, y, z) = (f.yup()(x, y+1, z) - f.ydown()(x, y-1, z))/(0.5*metric->dy(x, y+1) + metric->dy(x, y) + 0.5*metric->dy(x, y-1));
-      }
-  
-  for(int x=1;x<=mesh->LocalNx-2;x++) {
-    for(int y=mesh->ystart;y<=mesh->yend;y++) {
-      BoutReal by = 1./sqrt(metric->g_22(x, y));
-      for(int z=0;z<ncz;z++) {
-        // Z indices zm and zp
-        int zm = (z - 1 + ncz) % ncz;
-        int zp = (z + 1) % ncz;
-        
-        // bx = -DDZ(apar)
-        BoutReal bx = (apar(x, y, zm) - apar(x, y, zp))/(2.*metric->dz);
-        // bz = DDX(f)
-        BoutReal bz = (apar(x+1, y, z) - apar(x-1, y, z))/(0.5*metric->dx(x-1, y) + metric->dx(x, y) + 0.5*metric->dx(x+1, y));
-        
-	// Now calculate (bx*d/dx + by*d/dy + bz*d/dz) f
-        
-        // Length dl for predictor
-        BoutReal dl = fabs(metric->dx(x, y)) / (fabs(bx) + 1e-16);
-        dl = BOUTMIN(dl, fabs(metric->dy(x, y)) / (fabs(by) + 1e-16));
-        dl = BOUTMIN(dl, metric->dz / (fabs(bz) + 1e-16));
-        
-	BoutReal fp, fm;
-        
-        // X differencing
-        fp = f(x+1, y, z)
-          + (0.25*dl/metric->dz) * bz * (f(x+1, y, zm) - f(x+1, y, zp))
-          - 0.5*dl * by * gys(x+1, y, z);
-        
-        fm = f(x-1, y, z)
-          + (0.25*dl/metric->dz) * bz * (f(x-1, y, zm) - f(x-1, y, zp))
-          - 0.5*dl * by * gys(x-1, y, z);
-        
-        result(x, y, z) = bx * (fp - fm) / (0.5*metric->dx(x-1, y) + metric->dx(x, y) + 0.5*metric->dx(x+1, y));
-
-	// Z differencing
-        
-        fp = f(x, y, zp)
-          + (0.25*dl/metric->dx(x, y)) * bx * (f(x-1, y, zp) - f(x+1, y, zp))
-          - 0.5*dl * by * gys(x, y, zp);
-        
-        fm = f(x, y, zm)
-          + (0.25*dl/metric->dx(x, y)) * bx * (f(x-1,y,zm) - f(x+1, y, zm))
-          - 0.5*dl * by * gys(x, y, zm);
-
-        result(x, y, z) += bz * (fp - fm) / (2.*metric->dz);
-
-        // Y differencing
-        
-        fp = f.yup()(x,y+1,z)
-          - 0.5*dl * bx * (f.yup()(x+1, y+1, z) - f.yup()(x-1, y+1, z))/(0.5*metric->dx(x-1, y) + metric->dx(x, y) + 0.5*metric->dx(x+1, y))
-          
-          + (0.25*dl/metric->dz) * bz * (f.yup()(x,y+1,zm) - f.yup()(x,y+1,zp));
-        
-        fm = f.ydown()(x,y-1,z)
-          - 0.5*dl * bx * (f.ydown()(x+1, y-1, z) - f.ydown()(x-1, y-1, z))/(0.5*metric->dx(x-1, y) + metric->dx(x, y) + 0.5*metric->dx(x+1, y))
-          + (0.25*dl/metric->dz) * bz * (f.ydown()(x,y-1,zm) - f.ydown()(x,y-1,zp));
+  BOUT_FOR(i, f.getRegion("RGN_NOY")) {
```

Comment:
Note RGN_NOY isn't exactly what we had originally -- originally we calculate `gys` at all y apart for jy=0 and jy=LocalNy-1. We don't actually seem to use `gys` in the boundaries later though so it seem RGN_NOY is sufficient here?